src/octoprint/plugins/softwareupdate/__init__.py

```diff
@@ -1331,9 +1329,71 @@ def perform_update(self):
 
         force = json_data.get("force", "false") in octoprint.settings.valid_boolean_trues
 
+        if self._printer.is_printing() or self._printer.is_paused():
+            # do not update while a print job is running
+            # store targets to be run later on print done event
+            self._queued_updates["targets"] = list(
+                set(self._queued_updates["targets"] + targets)
+            )
+            self._send_client_message(
+                "queued_updates",
+                {"targets": self._queued_updates["targets"]},
+            )
+
+            return (
+                flask.jsonify({"queued": self._queued_updates.get("targets", False)}),
+                202,
+            )
+
         to_be_checked, checks = self.perform_updates(targets=targets, force=force)
         return flask.jsonify({"order": to_be_checked, "checks": checks})
 
+    @octoprint.plugin.BlueprintPlugin.route("/cancel_queued", methods=["POST"])
+    @no_firstrun_access
+    @Permissions.PLUGIN_SOFTWAREUPDATE_UPDATE.require(403)
+    def cancel_queued(self):
+        if "application/json" not in flask.request.headers["Content-Type"]:
+            flask.abort(400, description="Expected content-type JSON")
+
+        json_data = flask.request.get_json(silent=True)
+        if json_data is None:
+            flask.abort(400, description="Invalid JSON")
+
+        if "targets" in json_data:
+            targets = list(
+                map(
+                    lambda x: x.strip(),
+                    json_data.get("targets", []),
+                )
+            )
+        else:
+            targets = []
+
+        if len(targets) > 0:
+            if self._queued_updates:
+                for target in targets:
+                    if target in self._queued_updates.get("targets", []):
+                        self._queued_updates["targets"].remove(target)
+        else:
+            self._queued_updates["targets"] = []
+
+        if (
+            len(self._queued_updates.get("targets", [])) == 0
+            and self._queued_updates_abort_timer is not None
+        ):
+            self._queued_updates_abort_timer.cancel()
+            self._queued_updates_abort_timer = None
+
+        self._send_client_message(
+            "queued_updates",
+            {"targets": self._queued_updates["targets"]},
+        )
+
+        return (
+            flask.jsonify({"queued": self._queued_updates.get("targets", False)}),
+            202,
+        )
+
```

Comment:
This endpoint somehow feel icky still. And I'm talking a purely API design perspective here.

I think it would be better to have this endpoint be located at `/update/queued` and implement the command pattern also used in the `SimpleApiPlugin` mixin. So a `POST`, yes, but taking commands, only one for now, `cancel`, with optional parameter `targets`. The logic otherwise is sound.

Something like this maybe? (just written in review, untested & no pre-commit here)

```python
    @octoprint.plugin.BlueprintPlugin.route("/update/queued", methods=["POST"])
    @no_firstrun_access
    @Permissions.PLUGIN_SOFTWAREUPDATE_UPDATE.require(403)
    def cancel_queued(self):
        if "application/json" not in flask.request.headers["Content-Type"]:
            flask.abort(400, description="Expected content-type JSON")

        json_data = flask.request.get_json(silent=True)
        if json_data is None:
            flask.abort(400, description="Invalid JSON")
        
        command = json_data.get("command")
        if "command" == "cancel":
            targets = [x.strip() for x in json_data.get("targets", [])]

            if targets:
                self._queued_updates["targets"] = [x for x in self._queued_updates["targets"] if x not in targets]
            else:
                self._queued_updates["targets"] = []
                
            if (
                not self._queued_updates["targets"]
                and self._queued_updates_abort_timer
            ):
                self._queued_updates_abort_timer.cancel()
                self._queued_updates_abort_timer = None

            self._send_client_message(
               "queued_updates",
                {"targets": self._queued_updates["targets"]},
            )

            return (
                flask.jsonify({"queued": self._queued_updates["targets"]}),
                202,
            )

        else:
            flask.abort(400, description="Expected a valid command")
```

Also makes things a bit more pythonic.

Reply:
These changes were made. Totally makes sense and allows for expanding in the future if necessary.